File: 0100-same-tree/0100-same-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        //base case
        if(p==NULL && q==NULL) return true;
        if(p==NULL || q==NULL) return false;
        //meaning of this line is if(p==NULL || q==NULL) return false;
        //ki ya to p NUll honga pr q NULL nahi Honga aur vice versa
        //kyoki 2no NULL nahi ho skte ye apan ne upar vali line me check kr liya he already
        //is vali line me if(p==NULL && q==NULL) return true;

        if(p->val!=q->val) return false;

        bool lst = isSameTree(p->left,q->left); //lst = left sub tree 
        if(lst==false) return false;

        bool rst = isSameTree(p->right,q->right);
        if(rst==false) return false;

        return true;
    }
};

File: 0100-same-tree/0100-same-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        //base condition
        if(p==null && q==null) return true;
        if(p==null || q==null) return false;
        if(p.val != q.val) return false;

        boolean lst = isSameTree(p.left , q.left);
        if(lst==false) return false;

        boolean rst = isSameTree(p.right, q.right);
        if(rst == false) return false;

        return true;
    }
}